lib/datadog/tracing/contrib/aws/service/states.rb

```diff
@@ -16,21 +16,35 @@ def add_tags(span, params)
               execution_arn = params[:execution_arn]
               state_machine_account_id = ''
 
+              if state_machine_arn
+                span.set_tag(Aws::Ext::TAG_STATE_MACHINE_ARN, state_machine_arn)
+                # https://docs.aws.amazon.com/step-functions/latest/apireference/API_StartExecution.html#API_StartExecution_RequestSyntax:~:text=Required%3A%20No-,stateMachineArn,-The%20Amazon%20Resource
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>:10
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName:PROD>
+                # There are 3 patterns to cover and attempt to capture the `myStateMachineName`, it should always be in index 6 and account_id at index 4
+                parts = state_machine_arn.split(':')
+                state_machine_name ||= parts[6]
+                state_machine_account_id = parts[4]
+              end
+
               if execution_arn
-                # 'arn:aws:states:us-east-1:123456789012:execution:example-state-machine:example-execution'
+                span.set_tag(Aws::Ext::TAG_STATE_EXECUTION_ARN, execution_arn)
+                # express
+                # arn:aws:states:sa-east-1:123456789012:express:targetStateMachineName:1234:5678
+                # standard
+                # arn:aws:states:sa-east-1:123456789012:execution:targetStateMachineName:1234
                 parts = execution_arn.split(':')
-                state_machine_name = parts[-2]
+                state_machine_name = parts[6]
                 state_machine_account_id = parts[4]
```

Comment:
These two fields are being overwritten here. Is is intentional that `execution_arn` overrides `state_machine_arn`?

Reply:
Makes sense.
You can leave the code as is.
Can you add a comment explaining what you just mentioned above? ^ So in the future we can be confident when making changes to this method?

Reply:
comment might be unecessary now as I added .nil? to only set if it is nil

Reply:
and im not entirely confident with https://github.com/DataDog/dd-trace-rb/pull/3159#discussion_r1340345818